### PR TITLE
Support squashfs as compression algorithm

### LINKED ISSUE #6273: Support squashfs as compression algorithm
LXD already supports unpacking squashfs in a number of situations but it doesn't support using mksquashfs to generate squashfs based images or backups.

The existing compression logic for images and backups should be updated so that one can set the compression_algorithm to squashfs.

Consumption of squashfs images should already pretty much just work, consuming squashfs backups will need a few code updates too.

### PR DESCRIPTION
**Pull request by ulziibuyan**
Supposedly closes #6273

### CODE REVIEW

File: lxd/backup/backup.go [hunk 1]
```diff
@@ -58,6 +59,26 @@ func GetInfo(r io.ReadSeeker) (*Info, error) {
...
+			}
+			defer os.Remove(tempfile.Name())
+
+			// Write compressed data
+			_, err = io.Copy(tempfile, r)
+			if err != nil {
+				return nil, err
+			}
+
+			tempfile.Close()
```
Comment (stgraber) [inline lxd/backup/backup.go]: Should either be a defer or should also be called on io.Copy failure.

File: lxd/container.go [hunk 2]
```diff
@@ -316,16 +317,58 @@ func containerCreateFromBackup(s *state.State, info backup.Info, data io.ReadSee
...
+		}
+		defer os.Remove(tempfile.Name())
+
+		// Write the compressed data
+		_, err = io.Copy(tempfile, data)
+		if err != nil {
+			return nil, err
+		}
+
+		tempfile.Close()
```
Comment (stgraber) [inline lxd/container.go]: Should either be a defer or should also be called if io.Copy fails

File: lxd/container.go [hunk 3]
```diff
@@ -316,16 +317,58 @@ func containerCreateFromBackup(s *state.State, info backup.Info, data io.ReadSee
...
+		// Prepare to pass the temporary file as program argument
+		decomArgs := append(decomArgs, tempfile.Name())
+
+		// Create another temporary file to write the decompressed data
+		tarData, err := ioutil.TempFile("", "lxd_decompress_")
```
Comment (stgraber) [inline lxd/container.go]: Missing Close call on tarData

File: lxd/images.go [hunk 4]
```diff
@@ -152,7 +152,7 @@ func compressFile(compress string, infile io.Reader, outfile io.Writer) error {
 		defer os.Remove(tempfile.Name())
 
 		// Prepare 'tar2sqfs' arguments
-		args := []string{"tar2sqfs", "--no-skip", "--force",
+		args := []string{"tar2sqfs", "--force",
```
Comment (stgraber) [inline lxd/images.go]: This change doesn't appear related to this commit?

File: lxd/images.go [hunk 5]
```diff
@@ -851,6 +851,9 @@ func getImageMetadata(fname string) (*api.ImageMetadata, string, error) {
 
 	// Open the tarball
 	if len(unpacker) > 0 {
+		if algo == ".squashfs" {
```
Comment (stgraber) [inline lxd/images.go]: Could do with a comment like `// sqfs2tar can only read from a file` so we remember what this chunk is for :)

**Comment by stgraber** [comment]
@ulziibuyan ping

**Comment by ulziibuyan** [comment]
Made the changes as requested.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Thanks!
